File: dashboard/server/test/test_raw_message_handling.py
import pytest
import time
import os
from unittest.mock import MagicMock, patch

# Import components to test
from lib.can.interface import CANInterfaceWrapper

class TestRawMessageHandling:
    """Tests for raw message handling in the CAN interface."""

    @pytest.fixture
    def can_interface(self):
        """Create a test CAN interface with mock behavior."""
        # Create a mock implementation since we can't access real hardware
        with patch('lib.can.interface.has_can_hardware', False):
            interface = CANInterfaceWrapper(
                node_id=0x01,
                channel='vcan0',  # Use virtual CAN
            )
            yield interface

    def test_register_raw_handler(self, can_interface):
        """Test registration of raw message handlers."""
        # Create a mock handler function
        mock_handler = MagicMock()
        
        # Register the handler
        can_interface.register_raw_handler(0x700, mock_handler)
        
        # In mock mode, we can't really verify the handler was registered correctly,
        # but we can ensure the method executes without errors

    def test_send_raw_message(self, can_interface):
        """Test sending raw CAN messages."""
        # Create test data
        test_data = bytes([0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08])
        
        # Test sending
        result = can_interface.send_raw_message(0x700, test_data)
        
        # In mock mode, this should return success
        assert result is True, "send_raw_message should return True in mock mode"
        
    def test_raw_message_length_validation(self, can_interface):
        """Test validation of message length."""
        # Create oversized data (more than 8 bytes, which is invalid for CAN)
        oversized_data = bytes([i for i in range(10)])
        
        # Send message with oversized data
        result = can_interface.send_raw_message(0x700, oversized_data)
        
        # Should fail due to length validation
        assert result is False, "Oversized messages should be rejected"
        
    @pytest.mark.skipif(not os.environ.get('RUN_HARDWARE_TESTS'), 
                       reason="Hardware tests disabled")
    def test_integration_with_hardware(self):
        """
        Integration test with real hardware.
        
        This test is skipped by default and only runs when RUN_HARDWARE_TESTS
        environment variable is set, as it requires physical CAN hardware.
        """
        # Create real interface
        with patch('lib.can.interface.has_can_hardware', True):
            interface = CANInterfaceWrapper(
                node_id=0x01,
                channel='can0',  # Use real CAN
            )

            # Set up a mock to record received messages
            received_messages = []
            def on_raw_message(can_id, data, length):
                received_messages.append((can_id, data[:length]))
            
            # Register handler
            interface.register_raw_handler(0x700, on_raw_message)
            
            # Start processing
            interface.start_processing()
            
            try:
                # Send a test message
                test_data = bytes([0x01, 0x02, 0x03, 0x04, 0x05])
                interface.send_raw_message(0x700, test_data)
                
                # Wait for message to be processed
                start_time = time.time()
                timeout = 1.0  # 1 second timeout
                
                while time.time() - start_time < timeout and not received_messages:
                    time.sleep(0.01)
                    
                # Verify we received the message we sent
                assert len(received_messages) > 0, "No messages received"
                assert received_messages[0][0] == 0x700, "Unexpected CAN ID"
                assert received_messages[0][1] == test_data, "Message data mismatch"
                
            finally:
                # Clean up
                interface.stop_processing()

File: components/common/lib/CrossPlatformCAN/test/test_raw_message_handling.cpp
#include <gtest/gtest.h>
#include "ProtobufCANInterface.h"

// Mock raw handler for testing registerRawHandler
static void testRawHandler(uint32_t can_id, const uint8_t* data, uint8_t length) {
    // This would normally handle the raw message, but in the test we just need it to exist
}

TEST(ProtobufCANRawHandlingTest, RawHandlerRegistration) {
    ProtobufCANInterface interface(0x123);
    
    // Register a raw handler for animation data (CAN ID 0x700)
    interface.registerRawHandler(0x700, testRawHandler);
    
    // We can't test if the handler gets called directly in a unit test
    // since we can't send/receive real CAN messages, but we can verify
    // the interface doesn't crash when registering raw handlers
}

TEST(ProtobufCANRawHandlingTest, SendRawMessage) {
    ProtobufCANInterface interface(0x123);
    
    // Create test data (8 bytes max for CAN)
    uint8_t testData[8] = {0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08};
    
    // Test sending raw message
    // Note: We can't verify actual transmission in a unit test, just that the method handles the call
    EXPECT_FALSE(interface.sendRawMessage(0x700, testData, 8));
    
    // Without a real CAN interface, this should return false,
    // as the underlying CANInterface hasn't been properly initialized
}

// Test to verify that raw message handling follows the design constraints
TEST(ProtobufCANRawHandlingTest, RawMessageConstraints) {
    ProtobufCANInterface interface(0x123);
    
    // Test data at maximum length
    uint8_t maxData[8] = {0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08};
    
    // Send with maximum length
    EXPECT_FALSE(interface.sendRawMessage(0x700, maxData, 8));
    
    // Testing with different CAN IDs in the reserved range
    EXPECT_FALSE(interface.sendRawMessage(0x700, maxData, 4));  // Animation data
    EXPECT_FALSE(interface.sendRawMessage(0x701, maxData, 4));  // Reserved for future expansion
    EXPECT_FALSE(interface.sendRawMessage(0x70F, maxData, 4));  // Upper bound of reserved range
}
